whelktool/scripts/cleanups/2020/06/lxl-3187-tgm-relink-saogf-links.groovy

```diff
@@ -0,0 +1,78 @@
+/**
+ *
+ * See LXL-3187 for more information
+ */
+
+report = getReportWriter("report.txt")
+gmgpcPrefixSwe = "https://id.kb.se/term/gmgpc/swe/"
+saogfPrefix = "https://id.kb.se/term/saogf/"
+
+selectByCollection('auth') { data ->
+    def instance = data.graph[1]
+
+    if (!instance?.inScheme?.'@id'?.contains("https://id.kb.se/term/gmgpc")) {
+        return
+    }
+
+    def relations = ['broader', 'narrower', 'related']
+
+    boolean updated = false
+
+    relations.each { relation ->
+        if (!instance[relation]) {
+            return
+        }
+
+        //Objects containing @type, prefLabel and sameAs
+        def toRemove = instance[relation].findAll{ hasSaogfSameAs(it) }
+        def toAdd  = toRemove.collect{ ['@id' : gmgpcPrefixSwe + it.sameAs[0][ID].tokenize('/').last()] }
+
+        //Objects containing a single id key-value pair
+        def toRemoveSimple = instance[relation].findAll{ it[ID]?.contains(saogfPrefix)}
+        def toAddSimple = toRemoveSimple.collect { ['@id' : gmgpcPrefixSwe + it[ID].tokenize('/').last()] }
+
+        instance[relation].removeAll{ hasSaogfSameAs(it) }
+        instance[relation].removeAll{ it[ID]?.contains(saogfPrefix) }
+
+        //Add if linked object exists
+        (toAdd + toAddSimple).each {
+            if (exists(it)) {
+                instance[relation].add(it[ID])
```

Comment:
Unless I miss something, this turns the input reference objects (`{"@id": "https://..."}`) into just strings (`"https://..."`)? (I'd expect the code to be just `it` (or `[(ID): it[ID]]` if the input isn't a reference but a full description).)

Reply:
Good catch! This slipped in when adding the exists logic.